hq/app/logic/IamRemediation.scala

```diff
@@ -79,14 +84,73 @@ object IamRemediation extends Logging {
+        val keyPreviousActivities = remediationHistory.activityHistory.filter { activity =>
+          activity.problemCreationDate.withTimeAtStartOfDay == lastRotatedDate.withTimeAtStartOfDay()
+        }
+        keyPreviousActivities match {
+          case Nil =>
+            // there is no recent activity for the given access key, so return None.
+            None
+          case remediationActivities =>
+            // get the most recent remediation activity
+            Some(remediationActivities.maxBy { activity =>
+              Days.daysBetween(activity.problemCreationDate.withTimeAtStartOfDay(), activity.dateNotificationSent.withTimeAtStartOfDay()).getDays
+            })
+        }
+      case None =>
+        // we do not expect this case, because every IAM access key has a lastRotatedDate. SHQ's model needs to be changed to reflect this.
+        None
+    }
+  }
+
+  private[logic] def identifyRemediationOperation(mostRecentRemediationActivity: Option[IamRemediationActivity], now: DateTime,
+    userRemediationHistory: IamUserRemediationHistory): Option[RemediationOperation] =
+    mostRecentRemediationActivity match {
+      case None =>
+        // if there is no recent activity, then the required operation must be a Warning
+        Some(RemediationOperation(userRemediationHistory, Warning, OutdatedCredential, problemCreationDate = now))
+      case Some(mostRecentActivity) =>
+        mostRecentActivity.iamRemediationActivityType match {
+        case Warning if now.isAfter(mostRecentActivity.dateNotificationSent.plusDays(daysBetweenWarningAndFinalNotification - 1)) =>
+          // if the most recent activity is a Warning and the last notification was sent at least `Config.daysBetweenWarningAndFinalNotification` ago,
+          // the required operation is a FinalWarning
+          Some(RemediationOperation(userRemediationHistory, FinalWarning, mostRecentActivity.iamProblem, mostRecentActivity.problemCreationDate))
+        case FinalWarning if now.isAfter(mostRecentActivity.dateNotificationSent.plusDays(daysBetweenFinalNotificationAndRemediation - 1)) =>
+          // if the most recent activity is a FinalWarning and the last notification was sent at least `Config.daysBetweenFinalNotificationAndRemediation` ago,
+          // the required operation is Remediation
+          Some(RemediationOperation(userRemediationHistory, Remediation, mostRecentActivity.iamProblem, mostRecentActivity.problemCreationDate))
+        case Remediation =>
+          // if the most recent activity is Remediation, then we have entered into an edge case.
+          // The operation should still be Remediation, because the access key must not have been successfully disabled last time it was marked as Remediation.
+          Some(RemediationOperation(userRemediationHistory, Remediation, mostRecentActivity.iamProblem, mostRecentActivity.problemCreationDate))
+        case _ => None
```

Comment:
It's needed because whilst we've handled the three cases of the sealed trait, we haven't handled the else statements nestled in the Warning and FinalWarning cases.